src/main/kotlin/org/jitsi/nlj/transform/node/incoming/TccGeneratorNode.kt

```diff
@@ -92,55 +89,44 @@ class TccGeneratorNode(
      */
     private fun addPacket(tccSeqNum: Int, timestamp: Long, isMarked: Boolean, ssrc: Long) {
         synchronized(lock) {
-            if (packetArrivalTimes.ceilingKey(windowStartSeq) == null) {
```

Comment:
Pretty sure this code was a port of chrome logic

Reply:
The first change in the PR ("Track tcc seq numbers with a ROC") changes the keys of the map from wrapping 16-bit values to full ints extended with a ROC (so they are always ascending).

The second change (this commit) changes the semantics of `buildFeedback` in the corner case when we can not fit all received packets in one feedback packet. We used to only send the first packet, and now we send feedback for all packets, using multiple feedback packets if necessary. Since the map's keys always ascend this could be simplified to a loop over its entries and a final `clear`.

Reply:
Sounds good...is it worth adding a test to verify this?

Reply:
I don't quite follow all the logic of the original code, but I think part of the existing code was handling the case where packet reordering happens.

E.g. packets 1 and 3 arrive, and you report a loss on 2; and then packet 2 arrives, and because of the syntax of TCC to send a report on 2 you have to remember the information about 3.

Am I right that the existing code handled this case? Does the new code handle it equivalently?

Reply:
That's a good question. I don't understand the previous code either. Let me look at it more carefully. I don't think the new code reports loss at all.

Reply:
Not reporting on a packet (which is between two reported-on packets) counts as reporting loss, so that part should be fine.

Reply:
It looks like what Chrome does (in the spot where our current code has "TODO: chrome does something more advanced") is that whenever you add a packet to `packetArrivalTimes` and you're starting a new packet, rather than clear the map, it purges all records of packets older than some `back_window` (which seems to be 500 ms by default, but as always in Chrome this is a field trial param).  This lets it re-send old reports, as I described.

I don't think our current code accomplishes this.  I don't know if it needs to be added in this PR, if you don't want; I could do a separate PR for it after you're done.

Reply:
Aha. It looks like `RtcpFbTccPacket` handles adding "not received" entries when enconters a gap.



> because of the syntax of TCC to send a report on 2 you have to remember the information about 3.

Is this just to avoid signaling 3 as "not received"? Do I understand correctly that by clearing the map our existing code does the same as the PR?

Reply:
Right, exactly.  Yes, our existing code does the same thing as the PR.

I'll do a follow-up PR to behave the same way as Chrome; we don't need to block your PR on this.